### PR TITLE
initial, very basic implementation of graphite api render

### PR DESCRIPTION
**Pull request by AnderEnder**
* simple queries without functions and globs
* query parameters are taken from one of the source: form, json, or url
* from and until support only basic: relative(seconds, hours, days, weeks, years, minutes and months) and absolute(now, yesterday, unix-timestamp)
* responce is only json

### CODE REVIEW

File: graphite-api/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ParseError::RenderFormat => write!(f, "Format cannot be parsed"),
+            ParseError::Time => write!(f, "Time cannot be parsed"),
+            ParseError::SystemTimeError(s) => write!(f, "System time error: {}", s),
+            ParseError::ParseIntError(s) => write!(f, "{}", s),
+            ParseError::EmptyString => write!(f, "Can not parse empty string"),
```
Comment (andy128k) [inline graphite-api/src/error.rs]: Cannot

File: graphite-api/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+use std::convert::From;
+use std::error::Error;
+use std::fmt;
+use std::num::ParseIntError;
+use std::time::SystemTimeError;
+
+#[derive(Debug, PartialEq)]
+pub enum ParseError {
+    RenderFormat,
+    SystemTimeError(String),
```
Comment (andy128k) [inline graphite-api/src/error.rs]: Duration is lost. 
https://doc.rust-lang.org/src/std/time.rs.html#587-591

File: graphite-api/src/find.rs [hunk 3]
```diff
@@ -151,7 +151,10 @@ fn walk_tree(
         .filter_map(|entry| {
             let (local_path, local_file_type) = match entry {
                 Ok(rentry) => (
-                    rentry.path().strip_prefix(&full_path).map(|x| x.to_owned()),
+                    rentry
+                        .path()
+                        .strip_prefix(&full_path)
+                        .map(std::borrow::ToOwned::to_owned),
```
Comment (andy128k) [inline graphite-api/src/find.rs]: Meh... Closure is more readable, IMHO

File: graphite-api/src/render.rs [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RenderResponce {
+    entries: Vec<Option<RenderResponceEntry>>,
+}
+
+pub fn render_handler(state: State<Args>, params: RenderQuery) -> Result<HttpResponse, Error> {
+    let dir = &state.path;
+    let response: Vec<RenderResponceEntry> = params
+        .target
+        .iter()
```
Comment (andy128k) [inline graphite-api/src/render.rs]: into_iter()

File: graphite-api/src/render.rs [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+        .map(|x| RenderResponceEntry {
+            target: x.to_string(),
```
Comment (andy128k) [inline graphite-api/src/render.rs]: target: x,

File: graphite-api/src/render.rs [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+            datapoints: walk(&dir, x, &params).unwrap(),
```
Comment (andy128k) [inline graphite-api/src/render.rs]: expect

File: graphite-api/src/find.rs (hunk 3 above)
Comment (AnderEnder) [reply graphite-api/src/find.rs]: Hmm, clippy yells on it.

File: graphite-api/src/render.rs (hunk 4 above)
Comment (AnderEnder) [reply graphite-api/src/render.rs]: clippy considered that into_iter() should be replaced with .iter(), so I did it

Comment (andy128k) [reply graphite-api/src/render.rs]: iter instead of into_iter is a cause of extra allocation in the loop (`target: x.to_string()`)

Comment (AnderEnder) [reply graphite-api/src/render.rs]: and we received:
> borrow of moved value: `params`

Comment (AnderEnder) [reply graphite-api/src/render.rs]:
```
    let response: Vec<RenderResponceEntry> = params
        .target
        .clone()
        .into_iter()
        .map(|x| RenderResponceEntry {
            datapoints: walk(&dir, &x, &params).unwrap(),
            target: x,
        })
        .collect();
```
Is it better?

Comment (AnderEnder) [reply graphite-api/src/render.rs]: remade it

**Pull request merged**